News widget

brickahax opened the pull request:
**Issue:**
#110 

**Solution:**
Make a News Widget which displays a list of news article previews that link back to their original source. The list of news on display can be filtered by topic and region using the 2 dropdown menus in the component.

**Test Method:** 
Jest, Enzyme, Snapshot testing

**Risk:**
Public API keys

**Reviewed By:**
@jeva126 @xinyi98

brickahax committed:
Added base news widget

xinyi98 committed:
add news widget preview

brickahax committed:
Merge pull request #1 from jeva126/news-widget-base

Added base news widget

xinyi98 committed:
add news widget to the body

brickahax committed:
Merge pull request #2 from jeva126/news-widget-preview

Add News widget preview

jeva126 committed:
Added api functionality to news widget

News Widget can now pull headlines using the News API.
Styling not complete.

brickahax committed:
Edited CSS for article display and adjusted how loading message works

brickahax committed:
Merge pull request #3 from jeva126/news-widget-css

Edited CSS for article display and adjusted how loading message works

brickahax committed:
Added frontend testing for NewsWidget

TessaGush reviewed (approved):
Can't see any issues. Just need to resolve branch conflict

max-gurr commented on client/src/components/NewsWidget/ArticleItem.js:
```diff
@@ -0,0 +1,40 @@
+    };
+
+    if (article) {
+        console.log(article.props.title);
```
Can you remove these console.logs please

max-gurr commented on client/src/components/NewsWidget/newsAPI.js:
```diff
@@ -0,0 +1,12 @@
+const NEWS_API_KEY = "91e24cff34754f14aae4b9eaf858855d";
```
We should be storing API keys in a .env file, rather than in the code.

jeva126 replied:
I raised this point in the team meeting on Saturday and the consensus was to just leave it as is. Is it critical that we do it the way you suggest for the purpose of assignment 2 ?

max-gurr commented on client/src/components/NewsWidget/newsAPI.js:
```diff
@@ -0,0 +1,12 @@
+        `https://newsapi.org/v2/top-headlines?category=${topic}&country=${country}&sortBy=publishedAt&apiKey=${NEWS_API_KEY}`
+    );
+    const json = await response.json();
+    console.log("api json articles");
```
These console.logs should also be removed

max-gurr commented on client/src/components/NewsWidget/NewsWidget.test.js:
```diff
@@ -0,0 +1,27 @@
+test("dropdown default values load in", () => {
+    expect(component.find(Select).at(0).props().defaultValue).toEqual("general");
+    expect(component.find(Select).at(1).props().defaultValue).toEqual("nz");
+});
```
There should be something to test that the component properly tries to get newsarticles from the API, and ideally something to check that the right data gets returned. This way we can know if the endpoint becomes inactive

max-gurr reviewed (changes requested):
Looks good, Ive left a few comments about changes to keep coding conventions consistent, and also a suggestion for testing.

DanielDowling999 reviewed (approved):
I think the tests you've written so far are pretty good, but they might need some beefing up in the future. Right now the written tests don't seem to cover testing the functionality of the news display, but otherwise everything looks good to me.

jeva126 committed:
Removed console.logs and added a test

Test case tests that data is actually returned from the getArticles function

max-gurr commented on client/package.json:
```diff
@@ -1,4 +1,4 @@
-{
+ {
```
is this meant to be here?

max-gurr commented on client/src/components/NewsWidget/NewsWidget.test.js:
```diff
@@ -0,0 +1,34 @@
+
+test("the data is returned from the getArticles function (API endpoint works)", () => {
+    return getArticles("general", "nz").then((data) => {
+        expect(data.length).toBeGreaterThan(1);
```
It'd be good if you could add some extra expect statements here, basically just for whatever parts of the data you're accessing in the frontend code. Since you're relying on those parts being there, it's good to include them as test conditions

max-gurr reviewed (approved):
Everything looks good! I've left a couple suggestions, but they're not essential

jeva126 committed:
Added more detailed test

Test now checks the data type of each field of an arbitrary element in the array of articles is not undefined

jeva126 committed:
minor code tweak

jeva126 committed:
Merge branch 'main-group-5' into news-widget
